ui/src/sandbox.rs

```diff
@@ -571,6 +584,72 @@ fn build_execution_command(
     cmd
 }
 
+fn parse_json_output(output: Vec<u8>) -> Result<(String, String)> {
+    let mut composed_stderr_string = String::new();
+    let mut composed_stdout_string = String::new();
+
+    let metadata_stream = cargo_metadata::Message::parse_stream(&output[..]);
+
+    for msg in metadata_stream {
+        let message = msg.context(UnableToParseCargoOutputSnafu)?;
+
+        match message {
+            cargo_metadata::Message::TextLine(line) => {
+                composed_stdout_string.push_str(&(line + "\n"))
+            }
+
+            cargo_metadata::Message::CompilerMessage(cargo_metadata::CompilerMessage {
+                message,
+                ..
+            }) => {
+                composed_stderr_string.push_str(&parse_diagnostic(message, true));
+            }
+
+            _ => {}
+        }
+    }
+
+    Ok((composed_stdout_string, composed_stderr_string))
+}
+
+fn parse_diagnostic(
+    diagnostic: cargo_metadata::diagnostic::Diagnostic,
+    should_output_message: bool,
+) -> String {
```

Comment:
Instead of building a brand new string here, it'd be a bit more efficient to accept a `&mut String` argument and append to it directly.